content/blog/LFX-Mentorship-Blog.md

```diff
@@ -0,0 +1,425 @@
+
+Supported postprocessors and more about it: [postprocessors](https://docs.tremor.rs/artefacts/postprocessors/)
+
+Supported codecs, that can be configured in yam file can be found here: [codecs](https://docs.tremor.rs/artefacts/codecs/)
+
+![Tremor Dot Diagrm](/img/blog/LFX-Mentorship/dot-diagram.png)
+
+- Create a topic using the following `gcloud` command:
+
+```bash
+gcloud pubsub topics create <`topic-name`>
+```
+
+- Set the env variable
+
+```bash
+export GOOGLE_APPLICATION_CREDENTIALS="<path-to-service-token-json-file>"
+```
+
+- Command used to run tremor:
+
+```bash
+TREMOR_PATH=tremor-script/lib target/debug/tremor server run -f outbound.trickle inbound.trickle test.yaml | jq .
+```
+
+- After running tremor, create a subscription:
+
+```bash
+{"command": "create_subscription", "project": "<project-id>", "topic": "<topic-name>", "subscription": "<name-of-subscription>", "message_ordering": <`message-ordering`>}
+```
+
+*where:*
+
+***<`message-ordering`>** - can be set to `true` or `false` . To receive the messages in order, set the message ordering property on the subscription you receive messages from. Receiving messages in order might increase latency.*
+
+***<`project-id`>** - id of the GCP project*
+
+***<`topic-name`>** - Name of the Pub/Sub topic*
+
+- To send a message:
+
+```bash
+{"command": "send_message", "project":"<project-id>", "topic":"<topic-name>", "data": <`data`>, "ordering_key": "<ordering-key>"}
+```
+
+*where:*
+
+***<`data`>** - `json` message to be sent to the topic*
+
+***<`ordering-key`>** - If non-empty, identifies related messages for which publish order should be respected. If a Subscription has enable_message_ordering set to true, messages published with the same non-empty ordering_key value will be delivered to subscribers in the order in which they are received by the pub/sub system. All PubsubMessages published in a given PublishRequest must specify the same ordering_key value.*
+
+## Validation Testing for Pub/Sub
+
+For validating pub/sub, we need to make sure that there's guaranteed delivery and the messages are ordered when message_ordering is enabled for the subscription ie the sequence of messages received is preserved. So before sending the message to pub/sub, we add a field `count` in the event (json) that increments every time we send a message. This would be done in the outbound trickle file. To validate that all the messages are received in order, we have a validation logic in the inbound trickle file that checks if the difference between the `count` value of the current message and the previous message is one, the order is maintained.
+
+We have 3 different ways for setting up the outbound trickle file the outbound trickle - using a [transient wal](https://docs.tremor.rs/workshop/examples/20_transient_gd/), [persistent wal](https://docs.tremor.rs/workshop/examples/21_persistent_gd/) and no wal.
```

Comment:
I think the use of a wal could get a bit more explanation. Reading this I'm not sure as to why there are three variations of the pipeline

Reply:
It was to observe its behavior in these different cases as we have here in the docs - https://docs.tremor.rs/workshop/examples/21_persistent_gd/#insights

Reply:
I mean reading the article I'm not sure why suddenly a WAL pops up in the middle of it and then doesn't get mentioned again.

If I put myself in the shoes of someone reading this without the context of having been part of your mentorship the connection is missing.

It's a bit of like a movie introduces a side character, give them a backstory, introduces them to the hero and then the character gets never mentioned again :D the audience is confused. (your google work being the hero and WAL being the side character).

I think either the WAL plot needs an ending and each WAL option you introduce below needs its own little site plot with setup, actions, effects, and observations. Or the WAL just remains unmentioned as it's not essential to the story of your hero.

Reply:
Okay, right - I'll look into this👍